src/runtime_src/core/pcie/tools/cloud-daemon/mpd.cpp

```diff
@@ -86,73 +118,486 @@ static int connectMsd(pcieFunc& dev, std::string ip, uint16_t port, int id)
+				data, resp_len);
+			resp = data.get();
+			break;
+		}
+		case FIREWALL: {
+	    	if (!plugin_cbs.get_firewall_data) {
+        		ret = -ENOTSUP;
+				break;
+			}
+			std::shared_ptr<struct xcl_mig_ecc> data;
+			ret = (*plugin_cbs.get_firewall_data)(dev.getIndex(),
+				data, resp_len);
+			resp = data.get();
+			break;
+		}
+		case DNA: {
+	    	if (!plugin_cbs.get_dna_data) {
+        		ret = -ENOTSUP;
+				break;
+			}
+			std::shared_ptr<struct xcl_dna> data;
+			ret = (*plugin_cbs.get_dna_data)(dev.getIndex(),
+				data, resp_len);
+			resp = data.get();
+			break;
+		}
+		case SUBDEV: {
+	    	if (!plugin_cbs.get_subdev_data) {
+        		ret = -ENOTSUP;
+				break;
+			}
+			std::shared_ptr<void> data;
+			ret = (*plugin_cbs.get_subdev_data)(dev.getIndex(),
+				data, resp_len);
+			resp = data.get();
+			break;
+		}
+		default:
+			ret = -ENOTSUP;	   
+			break;
+		}
+
+		if (!ret) {
+			processed = std::make_shared<sw_msg>(resp, resp_len, orig->id(),
+				   MB_REQ_FLAG_RESPONSE);
+			dev.log(LOG_INFO, "mpd daemon: response %d sent", req->req);
+			return FOR_LOCAL;
+		}
+		break;
+	}
+	case MAILBOX_REQ_USER_PROBE: {//useful for aws plugin
+		struct mailbox_conn_resp resp = {0};
+		size_t resp_len = sizeof(struct mailbox_conn_resp);
+		resp.conn_flags |= MB_PEER_READY;
+		processed = std::make_shared<sw_msg>(&resp, resp_len, orig->id(),
+			MB_REQ_FLAG_RESPONSE);
+		dev.log(LOG_INFO, "mpd daemon: response %d sent", req->req);
+		return FOR_LOCAL;
+	}
+	case MAILBOX_REQ_LOCK_BITSTREAM: {//optional
+	    if (!plugin_cbs.lock_bitstream) {
+        	ret = -ENOTSUP;
+			break;
+		}
+		ret = (*plugin_cbs.lock_bitstream)(dev.getIndex());
+		break;
+	}
+	case MAILBOX_REQ_UNLOCK_BITSTREAM: { //optional
+	    if (!plugin_cbs.unlock_bitstream) {
+        	ret = -ENOTSUP;
+			break;
+		}
+		ret = (*plugin_cbs.unlock_bitstream)(dev.getIndex());
+		break;
+	}
+	case MAILBOX_REQ_HOT_RESET: {//optional
+	    if (!plugin_cbs.hot_reset) {
+        	ret = -ENOTSUP;
+			break;
+		}
+		ret = (*plugin_cbs.hot_reset)(dev.getIndex());
+		break;
+	}
+	case MAILBOX_REQ_RECLOCK: {//optional
+		struct xclmgmt_ioc_freqscaling *obj =
+			reinterpret_cast<struct xclmgmt_ioc_freqscaling *>(req->data);
+	    if (!plugin_cbs.hot_reset) {
+        	ret = -ENOTSUP;
+			break;
+		}
+		ret = (*plugin_cbs.reclock2)(dev.getIndex(), obj);
+		break;
+	}
+	default:
+	    break;
+	}
+out:	
+	processed = std::make_shared<sw_msg>(&ret, sizeof(ret), orig->id(),
+		MB_REQ_FLAG_RESPONSE);
+    dev.log(LOG_INFO, "mpd daemon: response %d sent ret = %d", req->req, ret);
+    return FOR_LOCAL;
+}
+
+/*
+ * Function to notify sofeware mailbox online/offline.
+ * This is usefull for aws. Since there is no mgmt, when xocl driver is loaded,
+ * and before the mpd daemon is running, sending MAILBOX_REQ_USER_PROBE msg
+ * will timeout and get no response, so there is no chance to know the card is
+ * ready. 
+ * A workaround is, when the mpd open/close the mailbox instance, a fake
```

Comment:
I think that this is by design that the mpd is behaving as xclmgmt when there is no msd/xclmgmt. This is not a workaround :-).

Reply:
remove "workaround"